### PR TITLE
[REVIEW] Port nvstrings Numeric Handling functions to cuDF Python/Cython

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR resolves: #3988 
- [x] Plumb cudf strings cpp api's with python
- [x] Add unit-tests for the same
- [x] Add functions docs
- [x] Dependency PR: #4244, will update/sync this PR once #4244 is merged.

### CODE REVIEW

File: python/cudf/cudf/_libxx/strings/char_types.pxd [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+
+from libcpp.memory cimport unique_ptr
+from cudf._libxx.cpp.column.column_view cimport column_view
+from cudf._libxx.cpp.column.column cimport column
+
+cdef extern from "cudf/strings/char_types/char_types.hpp" \
+        namespace "cudf::strings" nogil:
+
+    ctypedef enum string_character_types:
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/strings/char_types.pxd]: Lets change the underlying C++ to strongly type this enum class with `int32_t` underneath, and then define an additional `ctypedef` in the Cython here above the extern block for use with future Python Enums

Comment (galipremsagar) [reply python/cudf/cudf/_libxx/strings/char_types.pxd]: @kkraus14, Made the changes.

File: cpp/include/cudf/strings/char_types/char_types.hpp [hunk 2]
```diff
@@ -27,7 +27,7 @@ namespace strings
  * @brief Character type values.
  * These types can be or'd to check for any combination of types.
  */
-enum string_character_types {
+enum string_character_types : int32_t {
```
Comment (kkraus14) [inline cpp/include/cudf/strings/char_types/char_types.hpp]: We want to make this an `enum class`:
```suggestion
enum class string_character_types : int32_t {
```

File: python/cudf/cudf/_libxx/strings/char_types.pxd [hunk 3]
```diff
@@ -8,14 +8,18 @@ cdef extern from "cudf/strings/char_types/char_types.hpp" \
...
     ctypedef enum string_character_types:
-        DECIMAL = 1 << 0                              # binary 00000001
-        NUMERIC = 1 << 1                              # binary 00000010
-        DIGIT = 1 << 2                                # binary 00000100
-        ALPHA = 1 << 3                                # binary 00001000
-        SPACE = 1 << 4                                # binary 00010000
-        UPPER = 1 << 5                                # binary 00100000
-        LOWER = 1 << 6                                # binary 01000000
-        ALPHANUM = DECIMAL | NUMERIC | DIGIT | ALPHA  # binary 00001111
+        DECIMAL 'cudf::strings::string_character_types::DECIMAL' = 1 << 0
```
Comment (kkraus14) [inline python/cudf/cudf/_libxx/strings/char_types.pxd]: We shouldn't define the `= 1 << 0` after this, just defining the C string literal should give us the value we want.

**Comment by galipremsagar** [comment]
> CI is failing with `string_character_types` issue, please take a look

Yes, looking into it

File: cpp/include/cudf/strings/char_types/char_types.hpp (hunk 2 above)
Comment (kkraus14) [reply cpp/include/cudf/strings/char_types/char_types.hpp]: Connected offline and we use this enum to do `|` and `&` operations which you can't do on enum classes without implementing them, so leaving as an enum for now.

File: CHANGELOG.md [hunk 4]
```diff
@@ -122,6 +122,8 @@
 - PR #4282 Porting nvstrings conversion functions from new libcudf++ to Python/Cython
 - PR #4306 Use libcudf++ `unary.pyx` cast instead of legacy cast
 - PR #4305 Move gpuarrow.pyx and related libarrow_cuda files into _libxx
```
Comment (kkraus14) [inline CHANGELOG.md]:
```suggestion
- PR #4305 Move gpuarrow.pyx and related libarrow_cuda files into `_libxx`
```

**Pull request merged**